client/src/util/centroid.js

```diff
@@ -1,61 +1,118 @@
-
-  for (let i = 0, len = annoArray.length; i < len; i += 1) {
-    if (annoArray[i] === annoValue) {
-      centroid.x += layoutXArray[i];
-      centroid.y += layoutYArray[i];
-      centroid.size += 1;
-    }
-  }
-
-  if (centroid[2] !== 0) {
-    centroid.x /= centroid.size;
-    centroid.y /= centroid.size;
-  }
+/* 
+  calcMedianCentroid goes through a given metadata category
+  fetches each cell's coordinates grouping by category value.
 
-  return [centroid.x, centroid.y];
-};
+  It then calculates the median value and puts that in the array
 */
 
-const calcMedianCentroid = (world, annoName, annoValue, layoutDimNames) => {
-  const centroidX = [];
-  const centroidY = [];
-  let hasFinite = false;
+const calcMedianCentroid = (
+  obsAnnotations,
+  obsLayout,
+  categoryName,
+  layoutDimNames,
+  categoricalSelection,
+  schemaObsByName
+) => {
+  const categoryArray = obsAnnotations.col(categoryName).asArray();
+
+  const layoutXArray = obsLayout.col(layoutDimNames[0]).asArray();
+  const layoutYArray = obsLayout.col(layoutDimNames[1]).asArray();
+  const coordinates = new Map();
+
+  // Iterate over all the cells in the category
+  for (let i = 0, len = categoryArray.length; i < len; i += 1) {
+    const categoryValue = categoryArray[i];
+
+    // Get the index of the categoryValue within the category
+    // If the category is truncated and this value is removed,
+    // it will not be assigned a category value and will not be
+    // labeled on the graph
+    const categoryValueIndex = categoricalSelection[
+      categoryName
+    ].categoryValueIndices.get(categoryValue);
 
-  const annoArray = world.obsAnnotations.col(annoName).asArray();
-  const layoutXArray = world.obsLayout.col(layoutDimNames[0]).asArray();
-  const layoutYArray = world.obsLayout.col(layoutDimNames[1]).asArray();
+    // Check to see if the current category is a user created annotation
+    // if the user created this category, do not create a label for the `unassigned` value
+    const isUserAnno = schemaObsByName[categoryName].writable;
+
+    if (
+      categoryValueIndex !== undefined &&
+      !(isUserAnno && categoryValue === unassignedCategoryLabel)
+    ) {
+      // Get the number of cells which are in the category value
+      const numInCategoryValue =
+        categoricalSelection[categoryName].categoryValueCounts[
+          categoryValueIndex
+        ];
+
+      // Create/fetch the valueArray,
+      // which is what the key points to in the `coordinates` hashmap
+      const valueArray = coordinates.get(categoryValue) || [
+        false, // hasFinite
+        0, // index
+        new Float32Array(numInCategoryValue), // x coordinates
+        new Float32Array(numInCategoryValue) // y coordinates
+      ];
+      const index = valueArray[1];
+      let hasFinite = valueArray[0];
 
-  for (let i = 0, len = annoArray.length; i < len; i += 1) {
-    if (annoArray[i] === annoValue) {
       hasFinite =
```

Comment:
nit: this would be clearer (to me :-) if written as:

```
hasFinite = hasFinite || (Number.isFinite(layoutXArray[i]) && Number.isFinite(layoutYArray[i]))
```

Reply:
Added